### PR TITLE
Refactored to be backend-agnostic

### LINKED ISSUE #8: Stop using `GlResources`
Since gfx is truly abstracted from the device now, you don't need GlResources anywhere within your code. Please make it generic, and where the parameter is explicitly required - use a wildcard (`<_>`).

### PR DESCRIPTION
**Pull request by stjahns**
Closes https://github.com/PistonDevelopers/gfx-debug-draw/issues/8.
No assumption of OpenGL, except for a lack of other shaders..
Depends on https://github.com/gfx-rs/gfx-rs/pull/628

### CODE REVIEW

File: src/debug_renderer.rs [hunk 1]
```diff
@@ -29,17 +25,21 @@ impl FromError<ProgramError> for DebugRendererError {
     }
 }
 
-pub struct DebugRendererBuilder<'a> {
-    graphics: &'a mut Graphics<GlDevice>,
+pub struct DebugRendererBuilder<'a, D: Device>
```
Comment (kvark) [inline src/debug_renderer.rs:28]: Could be `R: Resources`

File: src/text_renderer.rs [hunk 2]
```diff
@@ -26,34 +26,32 @@ use gfx::batch::bind;
...
         bitmap_font: BitmapFont,
-        font_texture: TextureHandle<GlResources>,
-    ) -> Result<TextRenderer, ProgramError> {
+        font_texture: TextureHandle<D::Resources>,
+    ) -> Result<TextRenderer<D>, ProgramError> {
+
+        // TODO - Currently only have shaders for OpenGL, and no way to pick the right shader
+        // sources ...
 
         let program = match graphics.device.link_program(VERTEX_SRC.clone(), FRAGMENT_SRC.clone()) {
```
Comment (stjahns) [inline src/text_renderer.rs]: Is there already a mechanism or a pattern for choosing the proper shader sources?

File: src/line_renderer.rs [hunk 3]
```diff
@@ -17,25 +17,20 @@ use gfx::{
...
-    GlResources,
-};
-
 use utils::{grow_buffer, MAT4_ID};
 use std::marker::PhantomData;
 
-pub struct LineRenderer {
-    program: ProgramHandle<GlResources>,
+pub struct LineRenderer<D: Device> {
+    program: ProgramHandle<D::Resources>,
```
Comment (kvark) [inline src/line_renderer.rs:25]: Could be `<R: Resources>`

File: src/text_renderer.rs [hunk 4]
```diff
@@ -26,34 +26,32 @@ use gfx::batch::bind;
...
-    GlDevice,
-    GlResources,
-};
-
 use bitmap_font::BitmapFont;
 use utils::{grow_buffer, MAT4_ID};
 
-pub struct TextRenderer {
-    program: ProgramHandle<GlResources>,
+pub struct TextRenderer<D: Device> {
```
Comment (kvark) [inline src/text_renderer.rs:33]: Same here

File: src/debug_renderer.rs [hunk 5]
```diff
@@ -29,17 +25,21 @@ impl FromError<ProgramError> for DebugRendererError {
...
     bitmap_font: Option<BitmapFont>,
-    bitmap_font_texture: Option<TextureHandle<GlResources>>,
+    bitmap_font_texture: Option<TextureHandle<D::Resources>>,
 }
 
-impl<'a> DebugRendererBuilder<'a> {
+impl<'a, D: Device> DebugRendererBuilder<'a, D> {
 
-    pub fn new(graphics: &'a mut Graphics<GlDevice>, frame_size: [u32; 2]) -> DebugRendererBuilder {
+    pub fn new(graphics: &'a mut Graphics<D>, frame_size: [u32; 2]) -> DebugRendererBuilder<D> {
```
Comment (stjahns) [inline src/debug_renderer.rs:42]: How would I handle Graphics<D> with only R: Resources as a param?

File: src/text_renderer.rs (hunk 2 above)
Comment (kvark) [inline src/text_renderer.rs]: Yes, you can see gfx_graphics using the shader versions:
https://github.com/PistonDevelopers/gfx_graphics/blob/master/src/back_end.rs#L148-L157

File: src/debug_renderer.rs (hunk 5 above)
Comment (kvark) [inline src/debug_renderer.rs:42]: Sorry, I missed the `Graphics` member. `D` it is then.

File: src/text_renderer.rs (hunk 2 above)
Comment (stjahns) [inline src/text_renderer.rs]: thanks, I'll give it a try!

**Pull request merged**